plugin/src/main/java/org/gradle/testretry/internal/executer/RetryTestResultProcessor.java

```diff
@@ -207,7 +220,53 @@ private boolean currentRoundFailedTestsExceedsMaxFailures() {
     }
 
     public RoundResult getResult() {
-        return new RoundResult(currentRoundFailedTests, previousRoundFailedTests, lastRun(), hasRetryFilteredFailures);
+        return new RoundResult(currentRoundFailedTests,
+            cleanedUpFailedTestsOfPreviousRound(),
+            lastRun(),
+            hasRetryFilteredFailures
+        );
+    }
+
+    /**
+     * When running tests via the JUnit's suite engine or using {@code @Nested} test classes,
+     * Gradle 5.0 does not report the intermediate test class nodes. This leads to a problem when such
+     * test classes are configured to be retried on class-level, as we cannot properly remove them
+     * from the previous round of failed tests without a proper test event from Gradle.
+     * <p/>
+     * For Gradle 5.0, we manually remove all test classes with no registers test methods,
+     * if we saw the test class during this round. We assume here, that those entries are
+     * just here because of the missing event from Gradle for the intermediate node.
+     * <p/>
+     * For Gradle version 5.1 and above we don't do this, as we expect events for those intermediate
+     * nodes.
+     * <p/>
+     * This solution is not perfect but still allows users to use classRetry with Gradle 5
+     * together with the suite engine and/or nested test classes.
+     *
+     * @return cleaned up failed test names of previous round
+     */
+    private TestNames cleanedUpFailedTestsOfPreviousRound() {
+        boolean isGradle50 = GradleVersion.current().getBaseVersion().equals(GradleVersion.version("5.0"));
+
+        if (isGradle50 && !testClassesSeenInCurrentRound.isEmpty() || previousRoundFailedTests.hasClassesWithoutTestNames()) {
```

Comment:
❓ Shouldn't it be `isGradle50 && !testClassesSeenInCurrentRound.isEmpty() && previousRoundFailedTests.hasClassesWithoutTestNames()`? If I understand it correctly, we will enter this block in the following cases when it is not necessary:

* `isGradle50 = false && previousRoundFailedTests.hasClassesWithoutTestNames()` -> we will simply recreate the `previousRoundFailedTests`
* `isGradle50 = true && !testClassesSeenInCurrentRound.isEmpty() && hasClassesWithoutTestNames = false`. In this case, we will never enter the L257 branch and also simply recreate the `previousRoundFailedTests`.

Reply:
You're right.

I think I was playing with the idea to convert the condition from `!a && !b` to `!(a || b)` and didn't properly revert it.

FIXED (aad6d860)